File: src/models/Project.ts
import mongoose, { Schema, Document } from "mongoose";

export interface IProject extends Document {
  projectId: string; // On-chain ID
  clientPubkey: string;
  freelancerPubkey?: string;
  title: string;
  description: string;
  budget: number;
  status: "open" | "in_progress" | "completed" | "cancelled";
  deadline?: Date;
  descriptionHash?: string;
  createdAt: Date;
  updatedAt: Date;
}

const ProjectSchema: Schema = new Schema(
  {
    projectId: { type: String, required: true, unique: true, index: true },
    clientPubkey: { type: String, required: true },
    freelancerPubkey: { type: String },
    title: { type: String, required: true },
    description: { type: String, required: true },
    budget: { type: Number, required: true },
    status: {
      type: String,
      enum: ["open", "in_progress", "completed", "cancelled"],
      default: "open",
    },
    deadline: { type: Date },
    descriptionHash: { type: String },
  },
  { timestamps: true },
);

export default mongoose.model<IProject>("Project", ProjectSchema);

File: trustlancer_protocol/programs/solanahub_protocol/src/state/project.rs
use anchor_lang::prelude::*;

#[account]
#[derive(InitSpace)]
pub struct Project {
    pub project_id: u64,
    pub client: Pubkey,
    pub freelancer: Pubkey,
    #[max_len(128)]
    pub title: String,
    #[max_len(32)]
    pub description_hash: [u8; 32],
    pub total_budget: u64,
    pub milestone_count: u8,
    pub status: ProjectStatus,
    pub created_at: i64,
    pub deadline: i64,
    pub category: u8,
    pub bump: u8,
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, PartialEq, Eq, InitSpace)]
pub enum ProjectStatus {
    Open,
    Assigned,
    InProgress,
    Completed,
    Disputed,
    Cancelled,
}
